File: apps/android/core/client/src/main/kotlin/social/waddle/android/client/ExtensionCommands.kt
package social.waddle.android.client

import social.waddle.client.ffi.WaddleAdhocAction
import social.waddle.client.ffi.WaddleAdhocStatus
import social.waddle.client.ffi.WaddleExtensionCommand
import social.waddle.client.ffi.WaddleExtensionCommandForm
import social.waddle.client.ffi.WaddleExtensionCommandFormField
import social.waddle.client.ffi.WaddleExtensionCommandNote
import social.waddle.client.ffi.WaddleExtensionCommandResult
import social.waddle.client.ffi.WaddleExtensionCommandScope
import social.waddle.client.ffi.WaddleExtensionFieldType
import social.waddle.client.ffi.WaddleExtensionFormField
import social.waddle.client.ffi.WaddleExtensionNoteType

/**
 * Domain types of the `urn:waddle:extension:1` slash-command surface
 * (XEP-0050 ad-hoc commands on the extensions component), mirroring
 * the typed FFI records one-to-one so app code never touches the
 * generated bindings.
 */

/** `waddle#command_scope` — where a command may be offered. */
enum class ExtensionCommandScope { GLOBAL, CHANNEL }

/** One discovered extension command with its composer metadata. */
data class ExtensionCommand(
    val serviceJid: String,
    val node: String,
    val name: String,
    val scope: ExtensionCommandScope,
    /** Slash trigger (without the leading `/`), e.g. `poll`. */
    val composerPrefix: String? = null,
    /** Form field the composer may fill inline from trailing text. */
    val inlineField: String? = null,
    /** Whether a bare `/prefix` executes without opening the palette. */
    val composerExecute: Boolean = false,
)

/** XEP-0050 §3 `action` attribute. */
enum class ExtensionCommandAction { EXECUTE, CANCEL, NEXT, PREV, COMPLETE }

/** XEP-0050 §3 `status` attribute on responses. */
enum class ExtensionCommandStatus { EXECUTING, COMPLETED, CANCELED }

/** XEP-0004 §3.3 field types (unknown wire types map to TEXT_SINGLE). */
enum class ExtensionFieldType {
    BOOLEAN,
    FIXED,
    HIDDEN,
    JID_MULTI,
    JID_SINGLE,
    LIST_MULTI,
    LIST_SINGLE,
    TEXT_MULTI,
    TEXT_PRIVATE,
    TEXT_SINGLE,
}

/** One `<option/>` of a list field. */
data class ExtensionFieldOption(val label: String?, val value: String)

/** One `<field/>` of a command-response form. */
data class ExtensionCommandField(
    val name: String,
    val label: String?,
    val type: ExtensionFieldType,
    val required: Boolean,
    val options: List<ExtensionFieldOption>,
    val values: List<String>,
    /**
     * Forbidden-field defense (web `isForbiddenExtensionCommandField`
     * parity, computed by the Rust parser): `text-private` fields and
     * secret-named vars never render an input or submit.
     */
    val blocked: Boolean = false,
)

/** The XEP-0004 form of a command response. */
data class ExtensionCommandForm(
    val title: String?,
    val instructions: String?,
    val fields: List<ExtensionCommandField>,
)

/** XEP-0050 §3 `<note/>` severity. */
enum class ExtensionNoteType { INFO, WARN, ERROR }

/** One `<note/>` diagnostic from a command response. */
data class ExtensionCommandNote(val type: ExtensionNoteType, val value: String)

/** A parsed extension command response. */
data class ExtensionCommandResult(
    val status: ExtensionCommandStatus,
    /** XEP-0050 `sessionid`, threaded verbatim into follow-ups. */
    val sessionId: String?,
    /** Actions the service allows for the next stage. */
    val actions: List<ExtensionCommandAction>,
    val form: ExtensionCommandForm?,
    val notes: List<ExtensionCommandNote>,
)

/** One submitted field value pair. */
data class ExtensionSubmitField(val name: String, val values: List<String>)

/** One XEP-0050 form-stage submission. */
data class ExtensionCommandSubmission(
    val serviceJid: String,
    val node: String,
    /** Threaded verbatim from the previous stage's response. */
    val sessionId: String?,
    val fields: List<ExtensionSubmitField>,
    val action: ExtensionCommandAction,
    val roomJid: String?,
)

/**
 * Typed outcome of one invoke/submit round-trip. [Failed.detail]
 * carries the stanza-error text (or condition) when the server
 * refused; `null` for transport-level failures.
 */
sealed interface ExtensionCommandCall {
    data class Ok(val result: ExtensionCommandResult) : ExtensionCommandCall
    data class Failed(val detail: String?) : ExtensionCommandCall
}

// ── FFI mappings ─────────────────────────────────────────────────────

internal fun WaddleExtensionCommand.toDomain(): ExtensionCommand = ExtensionCommand(
    serviceJid = serviceJid,
    node = node,
    name = name,
    scope = when (scope) {
        WaddleExtensionCommandScope.GLOBAL -> ExtensionCommandScope.GLOBAL
        WaddleExtensionCommandScope.CHANNEL -> ExtensionCommandScope.CHANNEL
    },
    composerPrefix = composerPrefix,
    inlineField = inlineField,
    composerExecute = composerExecute,
)

internal fun WaddleExtensionCommandResult.toDomain(): ExtensionCommandResult = ExtensionCommandResult(
    status = when (status) {
        WaddleAdhocStatus.EXECUTING -> ExtensionCommandStatus.EXECUTING
        WaddleAdhocStatus.COMPLETED -> ExtensionCommandStatus.COMPLETED
        WaddleAdhocStatus.CANCELED -> ExtensionCommandStatus.CANCELED
    },
    sessionId = sessionId,
    actions = actions.map { it.toDomain() },
    form = form?.toDomain(),
    notes = notes.map { it.toDomain() },
)

private fun WaddleAdhocAction.toDomain(): ExtensionCommandAction = when (this) {
    WaddleAdhocAction.EXECUTE -> ExtensionCommandAction.EXECUTE
    WaddleAdhocAction.CANCEL -> ExtensionCommandAction.CANCEL
    WaddleAdhocAction.NEXT -> ExtensionCommandAction.NEXT
    WaddleAdhocAction.PREV -> ExtensionCommandAction.PREV
    WaddleAdhocAction.COMPLETE -> ExtensionCommandAction.COMPLETE
}

internal fun ExtensionCommandAction.toFfi(): WaddleAdhocAction = when (this) {
    ExtensionCommandAction.EXECUTE -> WaddleAdhocAction.EXECUTE
    ExtensionCommandAction.CANCEL -> WaddleAdhocAction.CANCEL
    ExtensionCommandAction.NEXT -> WaddleAdhocAction.NEXT
    ExtensionCommandAction.PREV -> WaddleAdhocAction.PREV
    ExtensionCommandAction.COMPLETE -> WaddleAdhocAction.COMPLETE
}

private fun WaddleExtensionCommandForm.toDomain(): ExtensionCommandForm = ExtensionCommandForm(
    title = title,
    instructions = instructions,
    fields = fields.map { it.toDomain() },
)

private fun WaddleExtensionCommandFormField.toDomain(): ExtensionCommandField = ExtensionCommandField(
    name = `var`,
    label = label,
    type = when (fieldType) {
        WaddleExtensionFieldType.BOOLEAN -> ExtensionFieldType.BOOLEAN
        WaddleExtensionFieldType.FIXED -> ExtensionFieldType.FIXED
        WaddleExtensionFieldType.HIDDEN -> ExtensionFieldType.HIDDEN
        WaddleExtensionFieldType.JID_MULTI -> ExtensionFieldType.JID_MULTI
        WaddleExtensionFieldType.JID_SINGLE -> ExtensionFieldType.JID_SINGLE
        WaddleExtensionFieldType.LIST_MULTI -> ExtensionFieldType.LIST_MULTI
        WaddleExtensionFieldType.LIST_SINGLE -> ExtensionFieldType.LIST_SINGLE
        WaddleExtensionFieldType.TEXT_MULTI -> ExtensionFieldType.TEXT_MULTI
        WaddleExtensionFieldType.TEXT_PRIVATE -> ExtensionFieldType.TEXT_PRIVATE
        WaddleExtensionFieldType.TEXT_SINGLE -> ExtensionFieldType.TEXT_SINGLE
    },
    required = required,
    options = options.map { ExtensionFieldOption(label = it.label, value = it.value) },
    values = values,
    blocked = blocked,
)

private fun WaddleExtensionCommandNote.toDomain(): ExtensionCommandNote = ExtensionCommandNote(
    type = when (noteType) {
        WaddleExtensionNoteType.INFO -> ExtensionNoteType.INFO
        WaddleExtensionNoteType.WARN -> ExtensionNoteType.WARN
        WaddleExtensionNoteType.ERROR -> ExtensionNoteType.ERROR
    },
    value = value,
)

internal fun ExtensionSubmitField.toFfi(): WaddleExtensionFormField =
    WaddleExtensionFormField(`var` = name, values = values)

File: server/crates/waddle-xmpp-client-ffi/src/extension_commands.rs
//! `urn:waddle:extension:1` extension-command exports (slash
//! commands over XEP-0050 ad-hoc commands).
//!
//! Mirrors the wasm chat client's dynamic discovery one-to-one:
//! nothing is hardcoded — the extension service is found via
//! `disco#items` on the account domain (with the conventional
//! `extensions.<domain>` fallback), the command list via
//! `disco#items` on the service with the XEP-0050 commands node, and
//! per-command composer metadata via `disco#info` extension forms.
//! The shared builders and parsers live in
//! `waddle_xmpp_client::extension_commands`; this file adapts the
//! typed uniffi Records and orchestrates the IQ round-trips.
//!
//! Discovery is cache-free by design: the Kotlin/Swift session layer
//! caches per session and clears on login/logout.

use jid::BareJid;
use minidom::Element;

use waddle_xmpp_client::discovery::{
    build_disco_info_iq, build_disco_items_iq, parse_disco_info_result, parse_disco_items_result,
};
use waddle_xmpp_client::extension_commands as core;
use waddle_xmpp_client::xep::xep0050::{AdHocAction, AdHocStatus};
use waddle_xmpp_client::{ClientError, ClientHandle};

use crate::boundary_convert::jid_domain;
use crate::error::client_error_to_waddle;
use crate::messaging_verbs::send_iq_with_timeout;
use crate::{
    WaddleAdhocAction, WaddleAdhocStatus, WaddleClient, WaddleError, WaddleExtensionCommand,
    WaddleExtensionCommandForm, WaddleExtensionCommandFormField, WaddleExtensionCommandNote,
    WaddleExtensionCommandResult, WaddleExtensionCommandScope, WaddleExtensionFieldOption,
    WaddleExtensionFieldType, WaddleExtensionFormField, WaddleExtensionNoteType,
};

// ── FFI ↔ core mappings ─────────────────────────────────────────────────────

pub(crate) fn adhoc_action_from_ffi(action: WaddleAdhocAction) -> AdHocAction {
    match action {
        WaddleAdhocAction::Execute => AdHocAction::Execute,
        WaddleAdhocAction::Cancel => AdHocAction::Cancel,
        WaddleAdhocAction::Next => AdHocAction::Next,
        WaddleAdhocAction::Prev => AdHocAction::Prev,
        WaddleAdhocAction::Complete => AdHocAction::Complete,
    }
}

fn adhoc_action_to_ffi(action: AdHocAction) -> WaddleAdhocAction {
    match action {
        AdHocAction::Execute => WaddleAdhocAction::Execute,
        AdHocAction::Cancel => WaddleAdhocAction::Cancel,
        AdHocAction::Next => WaddleAdhocAction::Next,
        AdHocAction::Prev => WaddleAdhocAction::Prev,
        AdHocAction::Complete => WaddleAdhocAction::Complete,
    }
}

fn adhoc_status_to_ffi(status: AdHocStatus) -> WaddleAdhocStatus {
    match status {
        AdHocStatus::Executing => WaddleAdhocStatus::Executing,
        AdHocStatus::Completed => WaddleAdhocStatus::Completed,
        AdHocStatus::Canceled => WaddleAdhocStatus::Canceled,
    }
}

fn scope_to_ffi(scope: core::ExtensionCommandScope) -> WaddleExtensionCommandScope {
    match scope {
        core::ExtensionCommandScope::Global => WaddleExtensionCommandScope::Global,
        core::ExtensionCommandScope::Channel => WaddleExtensionCommandScope::Channel,
    }
}

fn field_type_to_ffi(field_type: core::ExtensionFieldType) -> WaddleExtensionFieldType {
    match field_type {
        core::ExtensionFieldType::Boolean => WaddleExtensionFieldType::Boolean,
        core::ExtensionFieldType::Fixed => WaddleExtensionFieldType::Fixed,
        core::ExtensionFieldType::Hidden => WaddleExtensionFieldType::Hidden,
        core::ExtensionFieldType::JidMulti => WaddleExtensionFieldType::JidMulti,
        core::ExtensionFieldType::JidSingle => WaddleExtensionFieldType::JidSingle,
        core::ExtensionFieldType::ListMulti => WaddleExtensionFieldType::ListMulti,
        core::ExtensionFieldType::ListSingle => WaddleExtensionFieldType::ListSingle,
        core::ExtensionFieldType::TextMulti => WaddleExtensionFieldType::TextMulti,
        core::ExtensionFieldType::TextPrivate => WaddleExtensionFieldType::TextPrivate,
        core::ExtensionFieldType::TextSingle => WaddleExtensionFieldType::TextSingle,
    }
}

fn note_type_to_ffi(note_type: core::ExtensionNoteType) -> WaddleExtensionNoteType {
    match note_type {
        core::ExtensionNoteType::Info => WaddleExtensionNoteType::Info,
        core::ExtensionNoteType::Warn => WaddleExtensionNoteType::Warn,
        core::ExtensionNoteType::Error => WaddleExtensionNoteType::Error,
    }
}

fn command_to_ffi(descriptor: core::ExtensionCommandDescriptor) -> WaddleExtensionCommand {
    WaddleExtensionCommand {
        service_jid: descriptor.service_jid,
        node: descriptor.node,
        name: descriptor.name,
        scope: scope_to_ffi(descriptor.scope),
        composer_prefix: descriptor.composer_prefix,
        inline_field: descriptor.inline_field,
        composer_execute: descriptor.composer_execute,
    }
}

fn form_to_ffi(form: core::ExtensionCommandForm) -> WaddleExtensionCommandForm {
    WaddleExtensionCommandForm {
        title: form.title,
        instructions: form.instructions,
        fields: form
            .fields
            .into_iter()
            .map(|field| WaddleExtensionCommandFormField {
                var: field.var,
                label: field.label,
                field_type: field_type_to_ffi(field.field_type),
                required: field.required,
                blocked: field.blocked,
                options: field
                    .options
                    .into_iter()
                    .map(|option| WaddleExtensionFieldOption {
                        label: option.label,
                        value: option.value,
                    })
                    .collect(),
                values: field.values,
            })
            .collect(),
    }
}

fn result_to_ffi(result: core::ExtensionCommandResult) -> WaddleExtensionCommandResult {
    WaddleExtensionCommandResult {
        status: adhoc_status_to_ffi(result.status),
        session_id: result.session_id,
        actions: result
            .actions
            .into_iter()
            .map(adhoc_action_to_ffi)
            .collect(),
        form: result.form.map(form_to_ffi),
        notes: result
            .notes
            .into_iter()
            .map(|note| WaddleExtensionCommandNote {
                note_type: note_type_to_ffi(note.note_type),
                value: note.value,
            })
            .collect(),
    }
}

fn submit_fields_from_ffi(
    fields: Vec<WaddleExtensionFormField>,
) -> Vec<core::ExtensionSubmitField> {
    fields
        .into_iter()
        .map(|field| core::ExtensionSubmitField {
            var: field.var,
            values: field.values,
        })
        .collect()
}

// ── FFI → wire stanza factories ─────────────────────────────────────────────
//
// Pure functions between the FFI argument shapes and the shared
// builders; the dedicated test suite pins their output against the
// XEP wire shapes without a live connection.

pub(crate) fn extension_invoke_stanza(
    service_jid: &BareJid,
    node: &str,
    room_jid: Option<&BareJid>,
) -> Element {
    let room = room_jid.map(BareJid::to_string);
    core::build_extension_invoke_iq(&service_jid.to_string(), node, room.as_deref())
}

pub(crate) fn extension_submit_stanza(
    service_jid: &BareJid,
    node: &str,
    session_id: Option<&str>,
    action: AdHocAction,
    fields: &[core::ExtensionSubmitField],
    room_jid: Option<&BareJid>,
) -> Element {
    let room = room_jid.map(BareJid::to_string);
    core::build_extension_submit_iq(
        &service_jid.to_string(),
        node,
        session_id,
        action,
        fields,
        room.as_deref(),
    )
}

/// Map a command reply into the typed FFI result. A reply that lacks
/// the `<command/>` payload or carries an unknown status is a
/// server-side bug surfaced as [`WaddleError::MalformedResponse`].
pub(crate) fn map_command_reply(
    reply: Result<Element, ClientError>,
) -> Result<WaddleExtensionCommandResult, WaddleError> {
    let element = reply.map_err(|e| client_error_to_waddle(&e))?;
    core::parse_extension_command_result(&element)
        .map(result_to_ffi)
        .map_err(|_| WaddleError::MalformedResponse)
}

// ── Discovery orchestration ─────────────────────────────────────────────────

/// Resolve the extension service JID: probe every candidate with
/// `disco#info` and take the first advertising both the Waddle
/// extension feature and XEP-0050 command support; fall back to the
/// conventional `extensions.<domain>` when nothing qualifies.
async fn discover_extension_service(handle: &ClientHandle, domain: &str) -> String {
    let discovered = match send_iq_with_timeout(handle, build_disco_items_iq(domain, None)).await {
        Ok(reply) => parse_disco_items_result(&reply)
            .unwrap_or_default()
            .into_iter()
            .map(|item| item.jid)
            .collect(),
        Err(_) => Vec::new(),
    };
    for candidate in core::extension_service_candidates(domain, &discovered) {
        let Ok(reply) = send_iq_with_timeout(handle, build_disco_info_iq(&candidate, None)).await
        else {
            continue;
        };
        if parse_disco_info_result(&reply, &candidate)
            .is_some_and(|info| core::is_extension_service(&info))
        {
            return candidate;
        }
    }
    core::fallback_extension_service_jid(domain)
}

/// Fetch one command's composer metadata; a failed or unparsable
/// `disco#info` degrades to the metadata-free default (global scope,
/// no composer affordances) — the wasm client's fallback.
async fn fetch_command_metadata(
    handle: &ClientHandle,
    service_jid: &str,
    node: &str,
) -> core::ExtensionCommandMetadata {
    match send_iq_with_timeout(handle, build_disco_info_iq(service_jid, Some(node))).await {
        Ok(reply) => parse_disco_info_result(&reply, service_jid)
            .map(|info| core::parse_extension_command_metadata(&info))
            .unwrap_or_default(),
        Err(_) => core::ExtensionCommandMetadata::default(),
    }
}

// ── Exported verbs ───────────────────────────────────────────────────────────

#[uniffi::export(async_runtime = "tokio")]
impl WaddleClient {
    /// Discover the extension service and its XEP-0050 command list,
    /// hydrated with per-command composer metadata (scope, slash
    /// prefix, inline field, direct-execute). Cache-free: callers
    /// cache per session.
    pub async fn discover_extension_commands(
        &self,
    ) -> Result<Vec<WaddleExtensionCommand>, WaddleError> {
        let Some(handle) = self.clone_handle().await else {
            return Err(WaddleError::NotConnected);
        };
        let domain = jid_domain(&self.config.jid).to_string();
        let service_jid = discover_extension_service(&handle, &domain).await;
        let items_iq = build_disco_items_iq(
            &service_jid,
            Some(waddle_xmpp_client::xep::xep0050::NS_COMMANDS),
        );
        let reply = send_iq_with_timeout(&handle, items_iq)
            .await
            .map_err(|e| client_error_to_waddle(&e))?;
        let items = parse_disco_items_result(&reply).unwrap_or_default();
        let mut commands = Vec::new();
        for item in core::extension_command_refs(items, &service_jid) {
            let metadata = fetch_command_metadata(&handle, &item.service_jid, &item.node).await;
            commands.push(command_to_ffi(core::extension_command_descriptor(
                item, metadata,
            )));
        }
        Ok(commands)
    }

    /// XEP-0050 §2.4: start an extension command with `execute`. In a
    /// room, `room_jid` travels as the `waddle#room_jid` submit field.
    pub async fn invoke_extension_command(
        &self,
        service_jid: String,
        node: String,
        room_jid: Option<String>,
    ) -> Result<WaddleExtensionCommandResult, WaddleError> {
        let service: BareJid = service_jid.parse().map_err(|_| WaddleError::InvalidJid)?;
        let room = room_jid
            .map(|jid| jid.parse::<BareJid>())
            .transpose()
            .map_err(|_| WaddleError::InvalidJid)?;
        if node.trim().is_empty() {
            return Err(WaddleError::InvalidArgument);
        }
        let Some(handle) = self.clone_handle().await else {
            return Err(WaddleError::NotConnected);
        };
        let iq = extension_invoke_stanza(&service, &node, room.as_ref());
        map_command_reply(send_iq_with_timeout(&handle, iq).await)
    }

    /// XEP-0050 §3: submit a stage of an extension command session.
    /// `session_id` is threaded verbatim from the previous response;
    /// `cancel`/`prev` submissions never carry a form.
    pub async fn submit_extension_command_form(
        &self,
        service_jid: String,
        node: String,
        session_id: Option<String>,
        fields: Vec<WaddleExtensionFormField>,
        action: WaddleAdhocAction,
        room_jid: Option<String>,
    ) -> Result<WaddleExtensionCommandResult, WaddleError> {
        let service: BareJid = service_jid.parse().map_err(|_| WaddleError::InvalidJid)?;
        let room = room_jid
            .map(|jid| jid.parse::<BareJid>())
            .transpose()
            .map_err(|_| WaddleError::InvalidJid)?;
        if node.trim().is_empty() {
            return Err(WaddleError::InvalidArgument);
        }
        let Some(handle) = self.clone_handle().await else {
            return Err(WaddleError::NotConnected);
        };
        let iq = extension_submit_stanza(
            &service,
            &node,
            session_id.as_deref(),
            adhoc_action_from_ffi(action),
            &submit_fields_from_ffi(fields),
            room.as_ref(),
        );
        map_command_reply(send_iq_with_timeout(&handle, iq).await)
    }
}
